contrib/twitchtv/twirp/twirp.go

```diff
@@ -0,0 +1,134 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+package twirp
+
+import (
+	"context"
+	"math"
+	"net/http"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/twitchtv/twirp"
+)
+
+type contextKey int
+
+const (
+	twirpErrorKey contextKey = 0
+	httpSpanKey   contextKey = 1
+)
+
+// WrapServer wraps a http.Handler to add distributed tracing to a Twirp server.
+func WrapServer(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		opts := []tracer.StartSpanOption{
+			tracer.SpanType(ext.SpanTypeHTTP),
+			tracer.Tag(ext.HTTPMethod, r.Method),
+			tracer.Tag(ext.HTTPURL, r.URL.String()),
+		}
+		if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(r.Header)); err == nil {
+			opts = append(opts, tracer.ChildOf(spanctx))
+		}
+
+		span, ctx := tracer.StartSpanFromContext(r.Context(), "twirp.request", opts...)
+		defer span.Finish()
+
+		ctx = context.WithValue(ctx, httpSpanKey, span)
+		r = r.WithContext(ctx)
+		h.ServeHTTP(w, r)
+	})
+}
+
+// NewServerHooks creates the callback hooks for a twirp server to perform tracing.
+func NewServerHooks(opts ...Option) *twirp.ServerHooks {
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return &twirp.ServerHooks{
+		RequestReceived:  requestReceivedHook(cfg),
+		RequestRouted:    requestRoutedHook(cfg),
+		ResponsePrepared: responsePreparedHook(cfg),
+		ResponseSent:     responseSentHook(cfg),
+		Error:            errorHook(cfg),
+	}
+}
+
+func requestReceivedHook(cfg *config) func(context.Context) (context.Context, error) {
+	return func(ctx context.Context) (context.Context, error) {
+		pkg, ok := twirp.PackageName(ctx)
+		if !ok {
+			pkg = "unknown package"
+		}
+		svc, ok := twirp.ServiceName(ctx)
+		if !ok {
+			svc = "unknown service"
+		}
+		name := pkg + "." + svc
```

Comment:
Can we find something better here for the cases where pkg and svc are unknown? Won't it be weird to have:
```
unknown package.unknown service
```
Maybe we can make it use an underline? Is that better? 🤷‍♂ 
```
unknown_package.unknown_service
```

Reply:
This got changed to tags instead of automagic service naming.